### PR TITLE
libbpf-tools: add solisten

### PR DESCRIPTION
**Pull request by chenhengqi**
```shell
$ sudo solisten
PID     COMM             RET LAT(us)  BACKLOG PROTO PORT  ADDR                            
241011  netcat           0   12       1       TCPv4 1216  0.0.0.0                         
241158  nginx            0   14       511     TCPv4 80    0.0.0.0                         
241158  nginx            0   4        511     TCPv6 80    ::                              
241158  nginx            0   1        511     TCPv4 80    0.0.0.0                         
241158  nginx            0   1        511     TCPv6 80    ::
```

Signed-off-by: Hengqi Chen <chenhengqi@outlook.com>

### CODE REVIEW

File: libbpf-tools/solisten.bpf.c [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+/* Copyright (c) 2021 Hengqi Chen */
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_endian.h>
+#include <bpf/bpf_tracing.h>
+#include "solisten.h"
+
+#define MAX_ENTRIES	10240
+#define AF_INET		2
```
Comment (danobi) [inline libbpf-tools/solisten.bpf.c:11]: Weird indentation
```suggestion
#define AF_INET	2
```

File: libbpf-tools/solisten.c [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+	if (type == SOCK_STREAM)
+		prot = "TCP";
+	else if (type == SOCK_DGRAM)
+		prot = "UDP";
+	else
+		prot = "UNK";
+	snprintf(proto, sizeof(proto), "%s%s", prot, family == AF_INET ? "v4" : "v6");
+	inet_ntop(family, e->addr, addr, sizeof(addr));
+	printf("%-7d %-16s %-3d %-8d %-7d %-5s %-5d %-32s\n",
+	       e->pid, e->task, e->ret, (int)e->time / 1000, e->backlog, proto, e->port, addr);
```
Comment (danobi) [inline libbpf-tools/solisten.c:110]: Why cast `e->time` to int? Why not just use `%lu` format specifier?

File: libbpf-tools/solisten.bpf.c [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	return 0;
+}
+
+SEC("kprobe/inet_listen")
+int BPF_KPROBE(inet_listen_entry, struct socket *sock, int backlog)
+{
+	return probe_entry(sock, backlog);
+}
+
+SEC("kretprobe/inet_listen")
```
Comment (danobi) [inline libbpf-tools/solisten.bpf.c:82]: Can these 2 probes be replaced with a single `fexit/inet_listen` probe? Would let us remove the `values` map.

File: libbpf-tools/solisten.bpf.c [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	__u32 pid = pid_tgid >> 32;
+	__u32 tid = (__u32)pid_tgid;
+	__u16 family, type;
+	struct sock *sk = BPF_CORE_READ(sock, sk);
+	struct inet_sock *inet = (struct inet_sock *)sk;
+	struct event event = {
+		.time = bpf_ktime_get_ns(),
+		.pid = pid,
+		.backlog = backlog,
+	};
```
Comment (anakryiko) [inline libbpf-tools/solisten.bpf.c:41]: all this work before we even decide if we do the tracing or not (target_pid check below). Let's not be wasteful here

File: libbpf-tools/solisten.bpf.c [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+		return 0;
+
+	family = BPF_CORE_READ(sk, __sk_common.skc_family);
+	type = BPF_CORE_READ(sock, type);
+	event.proto = ((__u32)family << 16) | type;
+	event.port = bpf_ntohs(BPF_CORE_READ(inet, inet_sport));
+	if (family == AF_INET)
+		event.addr[0] = BPF_CORE_READ(sk, __sk_common.skc_rcv_saddr);
+	else if (family == AF_INET6)
+		bpf_probe_read_kernel(event.addr, sizeof(event.addr), sk->__sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32);
```
[the remark this answers is not in the document]
Comment (anakryiko) [reply libbpf-tools/solisten.bpf.c:53]: you could also do BPF_CORE_READ_INTO() here for consistency with BPF_CORE_READ() usage above

File: libbpf-tools/solisten.bpf.c [hunk 6]
```diff
@@ -0,0 +1,88 @@
+/* SPDX-License-Identifier: GPL-2.0 */
```
Comment (anakryiko) [inline libbpf-tools/solisten.bpf.c:1]: LGPL-2.1 OR BSD-2-Clause to match LICENSE variable below?

File: libbpf-tools/solisten.c [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+		return 1;
+	}
+
+	obj = solisten_bpf__open();
+	if (!obj) {
+		warn("failed to open BPF object\n");
+		return 1;
+	}
+
+	err = solisten_bpf__load(obj);
```
Comment (anakryiko) [inline libbpf-tools/solisten.c:146]: you can combine open and load into a single solisten_bpf__open_and_load(), given you are not doing any adjustments between those two steps

Comment (chenhengqi) [reply libbpf-tools/solisten.c:149]: I missed the `pid` filtering part. Added it.

File: libbpf-tools/solisten.bpf.c (hunk 3 above)
Comment (chenhengqi) [reply libbpf-tools/solisten.bpf.c:81]: Nice idea. Add `fexit` and fallback to `kprobe/kretprobe` if not available.

File: libbpf-tools/solisten.bpf.c [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+SEC("kprobe/inet_listen")
+int BPF_KPROBE(inet_listen_entry, struct socket *sock, int backlog)
+{
+	return probe_entry(sock, backlog);
+}
+
+SEC("kretprobe/inet_listen")
+int BPF_KRETPROBE(inet_listen_exit, int ret)
+{
+	return probe_exit(ctx, ret);
```
Comment (anakryiko) [inline libbpf-tools/solisten.bpf.c:90]: given probe_entry/probe_exit are not re-used in fexit program, I'd just inline them to have less need to jump around the source code to follow the logic

File: libbpf-tools/solisten.c [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+
+	obj = solisten_bpf__open();
+	if (!obj) {
+		warn("failed to open BPF object\n");
+		return 1;
+	}
+
+	obj->rodata->target_pid = target_pid;
+
+	if (vmlinux_btf_exists() && fentry_exists("inet_listen", NULL)) {
```
Comment (anakryiko) [inline libbpf-tools/solisten.c:151]: doesn't fentry_exist() return false if vmlinux BTF can't be found?

File: libbpf-tools/solisten.c [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+	err = solisten_bpf__attach(obj);
+	if (err) {
+		warn("failed to attach BPF programs: %d\n", err);
+		goto cleanup;
+	}
+
+	pb_opts.sample_cb = handle_event;
+	pb_opts.lost_cb = handle_lost_events;
+	pb = perf_buffer__new(bpf_map__fd(obj->maps.events), PERF_BUFFER_PAGES,
+			      &pb_opts);
```
Comment (anakryiko) [inline libbpf-tools/solisten.c:173]: nit: single line should be fine here

**Review by anakryiko — APPROVED** [review]
few minor nits, but otherwise looks good to me

**Comment by danobi** [comment]
Is there a concrete use case for the kprobe fallback? solisten already exists as a bcc-based tool; why not target the latest and greatest linux for a new tool?

The way I've seen fallback code like this go is it never, ever gets removed b/c of concerns over breaking existing users. But for new tools there are no existing users.

**Comment by chenhengqi** [comment]
> Is there a concrete use case for the kprobe fallback? solisten already exists as a bcc-based tool; why not target the latest and greatest linux for a new tool?
> 
> The way I've seen fallback code like this go is it never, ever gets removed b/c of concerns over breaking existing users. And then you're forever maintaining 2 of everything. But for new tools there are no existing users.

I saw first discussion here #3214  , and followed this way in #3441 and #3472 .

**Comment by anakryiko** [comment]
> Is there a concrete use case for the kprobe fallback? solisten already exists as a bcc-based tool; why not target the latest and greatest linux for a new tool?
> 

If we are trying to build a practical light-weight replacements for Python-based tools, we shouldn't just target the latest and greatest kernels, but rather try to make it work on as many version as possible without too much maintenance burden. You are right that we probably won't be able to remove kprobes any time soon, but overall I don't think it adds much of a problem.

**Comment by anakryiko** [comment]
Keep in mind that some embedded system might package and deploy libbpf-based variants, but never Python/BCC-based ones. Artificially crippling libbpf-based tools by chasing only the very latest features isn't great for end users.

This is one of the main reasons why I haven't tried forcing more of BPF ringbuf usage, for example.

**Comment by chenhengqi** [comment]
@danobi  Please take a look at this again.

**Review by danobi — APPROVED** [review]

**Comment by yonghong-song** [comment]
[buildbot, test this please]

**Pull request merged**